### PR TITLE
Convert API client to use AWS Go SDK

### PR DESCRIPTION
**Pull request by euank**
This replaces the generated client with the official AWS Go SDK's client.
This also improves testing around the api client as a bonus.

This will make it easier to update the client in the future, provides better connection pooling (via using http.Client directly), and hopefully does not have any major regressions.

Currently known regression: the '-k' flag does not work correctly with this. I'll add a commit that fixes that to this request before I merge it, but feel free to review this now.

### CODE REVIEW

File: agent/api/api_client.go [hunk 1]
```diff
@@ -14,158 +14,112 @@
...
-	return client.credentialProvider
+	client := ecs.New(ecsConfig)
+	return &ApiECSClient{credentialProvider: credentialProvider, config: config, insecureSkipVerify: insecureSkipVerify, c: client}
 }
 
-func getCpuAndMemory() (int32, int32) {
+func getCpuAndMemory() (int64, int64) {
 	memInfo, err := system.ReadMemInfo()
-	mem := int32(memInfo.MemTotal / 1024 / 1024) // MB
+	mem := memInfo.MemTotal / 1024 / 1024 // MB
```
Comment (aaithal) [inline agent/api/api_client.go]: Nit: This is MiB, not MB

File: agent/api/api_client.go [hunk 2]
```diff
@@ -14,158 +14,112 @@
...
+	"github.com/aws/amazon-ecs-agent/agent/httpclient"
 	"github.com/aws/amazon-ecs-agent/agent/logger"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
 var log = logger.ForModule("api client")
 
+// ECSClient is an interface over the ECSSDK interface which abstracts away some
+// details around constructing the request and reading the response down to the
+// parts ECS cares about.
```
Comment (samuelkarp) [inline agent/api/api_client.go]: s/ECS/the agent/

**Comment by samuelkarp** [comment]
:+1: 
This looks really good other than the two trivial comments above (mine and @aaithal's).

**Comment by euank** [comment]
@samuelkarp @aaithal I addressed your comments, corrected the retry logic, and plumbed through the -k flag.

The retry logic should be revisited again because the sdk should handle that correctly soon, but even if there's two layers of retries that's not all that big a deal.

**Pull request merged**